sql/handler.cc

```diff
@@ -8782,16 +8782,15 @@ bool handler::is_using_prohibited_gap_locks(TABLE* table,
   THD* thd = table->in_use;
   thr_lock_type lock_type = table->reginfo.lock_type;
 
-  if (!using_full_primary_key
-      && has_transactions()
-      && !has_gap_locks()
-      && thd_tx_isolation(thd) >= ISO_REPEATABLE_READ
-      && !thd->rli_slave
-      && (thd->lex->table_count >= 2 || thd->in_multi_stmt_transaction_mode())
-      && (lock_type >= TL_WRITE_ALLOW_WRITE ||
-          lock_type == TL_READ_WITH_SHARED_LOCKS ||
-          lock_type == TL_READ_NO_INSERT ||
-          (lock_type != TL_IGNORE && thd->lex->sql_command != SQLCOM_SELECT)))
+  if (!using_full_primary_key && has_transactions() && !has_gap_locks() &&
+      thd_tx_isolation(thd) >= ISO_REPEATABLE_READ && !thd->rli_slave &&
+      (thd->lex->table_count >= 2 || thd->in_multi_stmt_transaction_mode()) &&
+      (lock_type >= TL_WRITE_ALLOW_WRITE ||
+       lock_type == TL_READ_WITH_SHARED_LOCKS ||
+       lock_type == TL_READ_NO_INSERT ||
+       (lock_type != TL_IGNORE && thd->lex->sql_command != SQLCOM_SELECT)) &&
+      thd->lex->sql_command != SQLCOM_ALTER_TABLE &&
+      thd->lex->sql_command != SQLCOM_CHECK)
```

Comment:
@georgelorchpercona What would be the "permanent" fix for PS-4538? Can it be pushed separately from this PR?

Reply:
Reverting PS-3919 and fixing it correctly so that it does not change/create this new behavior that forces the gap lock detection logic to become involved for DDL/ANALYZE/OPTIMIZE ?  Otherwise this is the fix for PS-4538 and should probably be pushed on its own with specific test case(s).

Reply:
No opinion on the revert of PS-3919: on one hand, its benefit/regression ratio is extremely bad, on another, it's a 8.0 backport, so prepares us for things there

Reply:
The thing is, I don't believe we are seeing this same behavior (DDL triggering gap lock detection) in 8.0 (.4, .11, or .12).  So that implies that the backport may be incomplete or is possibly relying on behaviors found in 8.0 but not in 5.7.

Reply:
OK, I am correcting myself here, yes, I am now seeing this this behavior in Percona Server 8.0.  Consider this statement `alter table t modify h varchar(31) character set cp1257 collate cp1257_bin not null`.  In 5.7 this did NOT trigger a gap lock detection, in Percona Server 8.0 it does.  So it seems that proper fix here is to take this particular piece of this changeset into 5.7, either here or as a stand along change, and port it forward to 8.0.

Reply:
PS-3919 was reverted, this change should no longer be necessary.

Reply:
@vlad-lesin ^^ PS-3919 was reverted, this should no longer be needed and there should be no 'gap lock' related errors, please be sure you are re-based to include that commit https://github.com/Percona/percona-server/commit/4d803cdd69ec88a65e92a0b3fdadd75db4ce59c3

Reply:
Rebased on the newest 5.7, removed the above changes and tested: https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1933/ . The results look good to me.

Reply:
Yup, looks good.